File: backend/src/main/java/com/back/back9/domain/user/service/UserService.java
package com.back.back9.domain.user.service;

import com.back.back9.domain.user.dto.UserRegisterDto;
import com.back.back9.domain.user.entity.User;
import com.back.back9.domain.user.repository.UserRepository;
import com.back.back9.domain.wallet.service.WalletService;
import com.back.back9.global.exception.ServiceException;
import com.back.back9.global.rsData.RsData;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final AuthTokenService authTokenService;
    private final WalletService walletService;

    private RsData<User> registerUser(UserRegisterDto dto, User.UserRole role) {
        if (!dto.password().equals(dto.confirmPassword())) {
            return new RsData<>("400", "비밀번호가 일치하지 않습니다.");
        }
        if (userRepository.findByUserLoginId(dto.userLoginId()).isPresent()) {
            return new RsData<>("400-1", "이미 존재하는 아이디입니다.");
        }
        if (userRepository.findByUsername(dto.username()).isPresent()) {
            return new RsData<>("400-2", "이미 존재하는 유저이름입니다.");
        }

        String apiKey = UUID.randomUUID().toString();

        User user = User.builder()
                .userLoginId(dto.userLoginId())
                .username(dto.username())
                .password(passwordEncoder.encode(dto.password()))
                .role(role)
                .apiKey(apiKey)
                .build();

        user = userRepository.save(user);

        walletService.createWallet(user.getId());

        String message = role == User.UserRole.ADMIN ? "관리자 회원가입이 완료되었습니다." : "회원가입이 완료되었습니다.";
        return new RsData<>("200-1", message, user);
    }

    public RsData<User> register(UserRegisterDto dto) {
        return registerUser(dto, User.UserRole.MEMBER);
    }

    public RsData<User> registerAdmin(UserRegisterDto dto) {
        return registerUser(dto, User.UserRole.ADMIN);
    }

    public RsData<User> login(String userLoginId, String password) {
        Optional<User> userOpt = userRepository.findByUserLoginId(userLoginId);
        if (userOpt.isEmpty()) {
            return new RsData<>("401-1", "존재하지 않는 아이디입니다.");
        }
        User user = userOpt.get();
        if (!passwordEncoder.matches(password, user.getPassword())) {
            return new RsData<>("401-2", "비밀번호가 일치하지 않습니다.");
        }
        if (!walletService.existsByUserId(user.getId())) {
            walletService.createWallet(user.getId());
        }
        return new RsData<>("200-1", "로그인 성공", user);
    }

    public Optional<User> findByUserLoginId(String userLoginId) {
        return userRepository.findByUserLoginId(userLoginId);
    }

    public void deleteByUserLoginId(String userLoginId) {
        User user = userRepository.findByUserLoginId(userLoginId)
                .orElseThrow(() -> new ServiceException("404", "해당 아이디의 사용자가 없습니다."));
        walletService.deleteWalletByUserId(user.getId());
        userRepository.delete(user);
    }

    public Optional<User> findByApiKey(String apiKey) {
        return userRepository.findByApiKey(apiKey);
    }

    public List<User> searchByUsername(String keyword) {
        return userRepository.findByUsernameContaining(keyword);
    }

    public List<User> findAll() {
        return userRepository.findAll();
    }

    public Optional<User> findById(Long id) {
        return userRepository.findById(id);
    }

    public Map<String, Object> getPayloadFromToken(String accessToken) {
        return authTokenService.payload(accessToken);
    }

    public String genAccessToken(User user) {
        return authTokenService.genAccessToken(user);
    }

    public User save(User user) {
        return userRepository.save(user);
    }

    public void checkPassword(User user, String rawPassword) {
        if (!passwordEncoder.matches(rawPassword, user.getPassword())) {
            throw new ServiceException("401-1", "비밀번호가 일치하지 않습니다.");
        }
    }
}

File: backend/src/main/kotlin/com/back/back9/domain/user/service/UserService.kt
package com.back.back9.domain.user.service

import com.back.back9.domain.user.dto.UserRegisterDto
import com.back.back9.domain.user.entity.User
import com.back.back9.domain.user.repository.UserRepository
import com.back.back9.domain.wallet.service.WalletService
import com.back.back9.global.exception.ServiceException
import com.back.back9.global.rsData.RsData
import org.springframework.data.repository.findByIdOrNull
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class UserService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val authTokenService: AuthTokenService,
    private val walletService: WalletService
) {

    private fun registerUser(dto: UserRegisterDto, role: User.UserRole): RsData<User> {
        if (dto.password != dto.confirmPassword) {
            return RsData("400", "비밀번호가 일치하지 않습니다.")
        }
        if (userRepository.findByUserLoginId(dto.userLoginId) != null) {
            return RsData("400-1", "이미 존재하는 아이디입니다.")
        }
        if (userRepository.findByUsername(dto.username) != null) {
            return RsData("400-2", "이미 존재하는 유저이름입니다.")
        }

        val user = User(
            userLoginId = dto.userLoginId,
            username = dto.username,
            password = passwordEncoder.encode(dto.password),
            role = role,
            apiKey = UUID.randomUUID().toString()
        )

        val savedUser = userRepository.save(user)

        savedUser.id?.let { walletService.createWallet(it) }

        val message = if (role == User.UserRole.ADMIN) "관리자 회원가입이 완료되었습니다." else "회원가입이 완료되었습니다."
        return RsData("200-1", message, savedUser)
    }

    @Transactional
    fun register(dto: UserRegisterDto): RsData<User> {
        return registerUser(dto, User.UserRole.MEMBER)
    }

    @Transactional
    fun registerAdmin(dto: UserRegisterDto): RsData<User> {
        return registerUser(dto, User.UserRole.ADMIN)
    }

    @Transactional(readOnly = true)
    fun login(userLoginId: String, password: String): RsData<User> {
        val user = userRepository.findByUserLoginId(userLoginId)
            ?: return RsData("401-1", "존재하지 않는 아이디입니다.")

        if (!passwordEncoder.matches(password, user.password)) {
            return RsData("401-2", "비밀번호가 일치하지 않습니다.")
        }

        user.id?.let {
            if (!walletService.existsByUserId(it)) {
                walletService.createWallet(it)
            }
        }

        return RsData("200-1", "로그인 성공", user)
    }

    @Transactional(readOnly = true)
    fun findByUserLoginId(userLoginId: String): User? {
        return userRepository.findByUserLoginId(userLoginId)
    }

    @Transactional
    fun deleteByUserLoginId(userLoginId: String) {
        val user = userRepository.findByUserLoginId(userLoginId)
            ?: throw ServiceException("404", "해당 아이디의 사용자가 없습니다.")
        user.id?.let { walletService.deleteWalletByUserId(it) }
        userRepository.delete(user)
    }

    @Transactional(readOnly = true)
    fun findByApiKey(apiKey: String): User? {
        return userRepository.findByApiKey(apiKey)
    }

    @Transactional(readOnly = true)
    fun searchByUsername(keyword: String): List<User> {
        return userRepository.findByUsernameContaining(keyword)
    }

    @Transactional(readOnly = true)
    fun findAll(): List<User> {
        return userRepository.findAll()
    }

    @Transactional(readOnly = true)
    fun findById(id: Long): User? {
        return userRepository.findByIdOrNull(id)
    }

    fun getPayloadFromToken(accessToken: String): Map<String, Any> {
        return authTokenService.payload(accessToken)?.filterValues { it != null } as? Map<String, Any> ?: emptyMap()
    }

    fun genAccessToken(user: User): String {
        return authTokenService.genAccessToken(user)
    }

    fun save(user: User): User {
        return userRepository.save(user)
    }
}
